packages/@aws-cdk/core/README.md

```diff
@@ -1302,4 +1302,118 @@ permissions boundary attached.
 
 For more details see the [Permissions Boundary](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_iam-readme.html#permissions-boundaries) section in the IAM guide.
 
+## Policy Validation
+
+If you or your organization use any policy validation tool, such as
+[CloudFormation
+Guard](https://docs.aws.amazon.com/cfn-guard/latest/ug/what-is-guard.html) or
+[OPA](https://www.openpolicyagent.org/), to define constraints on your
+CloudFormation template, you can integrate them with the CDK at synthesis time.
+By using the appropriate plugin, you can make the CDK application check the
+generated CloudFormation template against your policies immediately after
+synthesis. If there are any violations, the synthesis will fail and a report
+will be printed to the console.
+
+### For application developers
+
+To use one or more validation plugins in your application, use the
+`validationPlugins` property of `Stage`:
+
+```ts
+// globally for the entire app (an app is a stage)
+const app = new App({
+  validationPlugins: [
+    // These hypothetical classes implement IValidationPlugin:
+    new ThirdPartyPluginX(), 
+    new ThirdPartyPluginY(),
+  ],
+});
+
+// only apply to a particular stage
+const prodStage = new Stage(app, 'ProdStage', {
+  validationPlugins: [...],
+});
+```
+
+Immediately after synthesis, all plugins registered this way will be invoked to
+validate all the templates generated in the scope you defined. In particular, if
+you register the templates in the `App` object, all templates will be subject to
+validation.
+
+By default, the report will be printed in a human readable format. If you want a
+report in JSON format, use the `synth()` method:
+
+```ts
+app.synth({
+  validationReportFormat: ValidationFormat.JSON
+});
+```
+
+### For plugin authors
+
+The communication protocol between the CDK core module and your policy tool is
+defined by the `IValidationPlugin` interface. To create a new plugin you must
+write a class that implements this interface. There are three things you need to
+implement: the plugin name (by overriding the `name` property), and the two
+methods `isReady()` and `validate()`.
+
+The method `isReady()` is called first in the workflow, to make sure that the
+plugin is in a valid state and can be used. For example, most plugins will have
+an external dependency, such as a CLI, which must be installed for the plugin to
+work. To signal to the framework that it can go ahead and use the plugin, you
+can check whether the CLI is installed by checking the version:
+
+```ts
+declare function invokeCliVersionCommand(): number;
+
+isReady(): boolean {
+  const status = invokeCliVersionCommand();
```

Comment:
Are these two functions expected to be synchronous? If so, will the CDK provide helpers for plugins authors to correctly synchronously invoke external executables?

Reply:
Yes. The whole framework is synchronous by design. This methods has to be synchronous as well.

Reply:
That makes sense and was my assumption. Would be awesome when this released and documented to have recommended patterns for safely and correctly invoking executables. I my (admittedly personal) experience, most devs are used to using the async versions of process invocation (`spawn`, `execa`) and might trip over this requirement at first.